File: leet-code/sortString.go
package main

import (
	"strings"
)

func sortString(s string) string {
	if len(s) == 0 {
		return ""
	}
	var bucket [26]int
	res := make([]string, 0)
	for i := range s {
		bucket[s[i]-'a']++
	}

	for len(res) < len(s) {
		for i := 0; i <= 25; i++ {
			if bucket[i] <= 0 {
				continue
			}
			res = append(res, string(rune(i)+'a'))
			bucket[i]--
		}

		for i := 25; i >= 0; i-- {
			if bucket[i] <= 0 {
				continue
			}
			res = append(res, string(rune(i)+'a'))
			bucket[i]--
		}
	}
	return strings.Join(res, "")
}

func sortString2(s string) string {
	if len(s) == 0 {
		return ""
	}
	bucket := [26]int{}
	for _, v := range s {
		bucket[v-'a']++
	}

	res := strings.Builder{}
	for res.Len() < len(s) {
		for c := 'a'; c <= 'z'; c++ {
			if bucket[c-'a'] > 0 {
				res.WriteRune(c)
				bucket[c-'a']--
			}
		}

		for c := 'z'; c >= 'a'; c-- {
			if bucket[c-'a'] > 0 {
				res.WriteRune(c)
				bucket[c-'a']--
			}
		}
	}
	return res.String()
}

File: leet-code/sortString.js
/**
 * @param {string} s
 * @return {string}
 */
var sortString = function (s) {
  if (s.length === 0) {
    return ''
  }
  const res = []
  // 字典
  const bucket = new Array(26).fill(0)
  for (let i = 0, len = s.length; i < len; i++) {
    bucket[s[i].codePointAt() - 'a'.codePointAt()]++
  }
  // 来回遍历,因为一定会排完，所以终止条件用res.length < s.length应该没啥问题。。
  while (res.length < s.length) {
    for (let i = 0; i <= 25; i++) {
      if (bucket[i] <= 0) {
        continue
      }
      res.push(String.fromCharCode(i + 97))
      bucket[i]--
    }
    for (let i = 25; i >= 0; i--) {
      if (bucket[i] <= 0) {
        continue
      }
      res.push(String.fromCharCode(i + 97))
      bucket[i]--
    }
  }
  return res.join('')
}

console.log(sortString('rat'))
